### DISCUSSION TITLE
OpenTelemetry: Custom sampler config; rate-limiting sampler

### DISCUSSION THREAD

**Discussion by calohmn** [quarkusio/quarkus#26438] [Q&A] [answered]
When migrating from the OpenTracing Jaeger client to using `quarkus-opentelemetry` with the OpenTracing shim, we want to keep the possibility of letting a **rate-limiting** sampler be configured (see [Jaeger client configuration](https://www.jaegertracing.io/docs/1.35/sampling/#client-sampling-configuration)).

With `quarkus-opentelemetry`, this currently isn't supported out-of-the-box. Possible values for [quarkus.opentelemetry.tracer.sampler](https://quarkus.io/guides/opentelemetry#quarkus-opentelemetry_quarkus.opentelemetry.tracer.sampler) are `off, on, ratio`, meaning either a constant (on/off) or probabilistic sampler can be used.

Of course a rate-limiting sampler can easily be integrated in a Quarkus application by providing a custom `Sampler` CDI bean.
But then the standard `quarkus.opentelemetry.tracer.sampler` configuration properties can't be used. I will have to think about some other configuration structure with which I allow my Quarkus application to use either an On/Off Sampler or a rate-limiting sampler for example; the rate-limiting sampler also needing a rate-limit configuration value.

Therefore:
- Would it be feasible to add support for a rate-limiting sampler as one of the "quarkus.opentelemetry.tracer.sampler" options?
I.e. having `off, on, ratio, rate-limiting` as possible values there. The sampler implementation would be straightforward, using the Opentelemetry SDK `RateLimiter` (see [here](https://github.com/open-telemetry/opentelemetry-java/blob/0860b3884b235274f50566908bb1b5849aaa6487/sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/RateLimitingSampler.java)). This feature would make it easier to migrate from the Jaeger client.

- And/or: Could a new `quarkus.opentelemetry.tracer.sampler.custom` configuration property be added - with value of type `Map<String,String>`?
This would allow arbitrary configuration properties under the standard `quarkus.opentelemetry.tracer.sampler.` prefix to be used by a custom Sampler. This would make it easier to add a custom Sampler that needs configuration.

I would appreciate feedback here on whether to create a feature request for one and/or the other option.

[a comment by an automated account is not shown]

**Reply by radcortez** [reply]
The plan is to support OTel Configuration OOTB. In that case, you would be able to set up the Sampler like this:
https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md#sampler

Unfortunately, I cannot give you a timeline of when that work is going to be completed. It is not started yet. It is in my mind to implement for a couple of months now, but I keep getting pulled to other stuff :)

**Reply by radcortez** [reply]
BTW the Jaeger `RateLimiter` `Sampler` is not listed in the standard configuration options, but it is registered as an SPI, so you would use `jaeger_remote` to configure it.

**Reply by calohmn** [reply]
> The plan is to support OTel Configuration OOTB.

Ah, good to know. 
So, that means the actual `otel.traces.sampler` and `otel.traces.sampler.arg` (system/env) properties shall be used then?
Or will this kind of sampler configuration be mapped to the existing `quarkus.opentelemetry.tracer.sampler` and an additional `quarkus.opentelemetry.tracer.sampler.arg` property?

> BTW the Jaeger `RateLimiter` `Sampler` is not listed in the standard configuration options, but it is registered as an SPI, so you would use `jaeger_remote` to configure it.

Good to know about the SPI mechanism here.
`jaeger_remote` wouldn't fit though, as I wouldn't want to configure fetching the sampling config from a remote Jaeger server. And the `RateLimitingSampler` class alongside the `JaegerRemoteSampler` isn't public.
But I could easily use my own `RateLimitingSamplerProvider` and `RateLimitingSampler`, I guess.

**Reply by radcortez** [reply]
> > The plan is to support OTel Configuration OOTB.
> 
> Ah, good to know. So, that means the actual `otel.traces.sampler` and `otel.traces.sampler.arg` (system/env) properties shall be used then? Or will this kind of sampler configuration be mapped to the existing `quarkus.opentelemetry.tracer.sampler` and an additional `quarkus.opentelemetry.tracer.sampler.arg` property?
> 

We usually wrap config integrations in our own namespace, so we can benefit from all our configuration mechanisms. I guess we will do something similar here, but most likely we will need to change some names to make the mapping more consistent. For instance `otel.traces.sampler` should become `quarkus.otel.traces.sampler` and not `quarkus.opentelemetry.tracer.sampler`. Anyway, there is no final decision on this yet.

> > BTW the Jaeger `RateLimiter` `Sampler` is not listed in the standard configuration options, but it is registered as an SPI, so you would use `jaeger_remote` to configure it.
> 
> Good to know about the SPI mechanism here. `jaeger_remote` wouldn't fit though, as I wouldn't want to configure fetching the sampling config from a remote Jaeger server. And the `RateLimitingSampler` class alongside the `JaegerRemoteSampler` isn't public. But I could easily use my own `RateLimitingSamplerProvider` and `RateLimitingSampler`, I guess.

I've mentioned that one because that is the one that links with the `RateLimiter` that you mentioned :) Maybe they have more options available (I haven't checked).

**Reply by calohmn** [reply]
Ok, thanks for the input. Looking forward to these changes then.

**Reply by calohmn** [reply]
Speaking of the sampler config property names:
When choosing a new property name for the *sampler name*, I think it should end with ".name" or ".id" then.

I mean, with the current
`quarkus.opentelemetry.tracer.sampler` and
`quarkus.opentelemetry.tracer.sampler.ratio` 
properties, I don't see how the sampler configuration can be done in a YAML file (along with other quarkus properties), as the `sampler` field would need both a string and a map value.

**Reply by radcortez** [reply]
Check here: https://quarkus.io/guides/config-yaml#configuration-key-conflicts
